File: bot-api/dotnet/src/Bot.cs
using System;
using Robocode.TankRoyale.BotApi.Events;
using Robocode.TankRoyale.BotApi.Internal;

namespace Robocode.TankRoyale.BotApi
{
  /// <summary>
  /// Abstract bot class provides convenient methods for movement, turning, and firing the gun.
  /// Most bots should inherit from this class.
  /// </summary>
  public abstract class Bot : BaseBot, IBot
  {
    private readonly BotInternals __botInternals;

    // <inheritdoc/> does not work with the default constructor?
    /// <summary>
    /// Constructor for initializing a new instance of the Bot class, which should be used when
    /// both BotInfo and server URI is provided through environment variables, i.e., when starting
    /// up the bot using a booter. These environment variables must be set to provide the server
    /// URL and bot information, and are automatically set by the booter tool for Robocode.
    ///
    /// Example of how to set the predefined environment variables:
    ///
    /// SERVER_URL=ws://localhost<br/>
    /// SERVER_SECRET=xzoEeVbnBe5TGjCny0R1yQ<br>
    /// BOT_NAME=MyBot<br/>
    /// BOT_VERSION=1.0<br/>
    /// BOT_AUTHOR=fnl<br/>
    /// BOT_DESCRIPTION=Sample bot<br/>
    /// BOT_URL=https://mybot.somewhere.net<br/>
    /// BOT_COUNTRY_CODE=DK<br/>
    /// BOT_GAME_TYPES=melee,1v1<br/>
    /// BOT_PLATFORM=.Net Core 3.1<br/>
    /// BOT_PROG_LANG=C# 8<br/>
    /// </summary>
    public Bot() : base()
    {
      __botInternals = new BotInternals(this, base.__baseBotInternals);
    }

    /// <inheritdoc/>
    public Bot(BotInfo botInfo) : base(botInfo)
    {
      __botInternals = new BotInternals(this, base.__baseBotInternals);
    }

    /// <inheritdoc/>
    public Bot(BotInfo botInfo, Uri serverUrl) : base(botInfo, serverUrl)
    {
      __botInternals = new BotInternals(this, base.__baseBotInternals);
    }

    /// <inheritdoc/>
    public virtual void Run() { }

    /// <inheritdoc/>
    public bool IsRunning => __botInternals.IsRunning;

    /// <inheritdoc/>
    public void SetForward(double distance)
    {
      __botInternals.SetForward(distance);
    }

    /// <inheritdoc/>
    public void Forward(double distance)
    {
      __botInternals.Forward(distance);
    }

    /// <inheritdoc/>
    public void SetBack(double distance)
    {
      SetForward(-distance);
    }

    /// <inheritdoc/>
    public void Back(double distance)
    {
      Forward(-distance);
    }

    /// <inheritdoc/>
    public double DistanceRemaining => __botInternals.DistanceRemaining;

    /// <inheritdoc/>
    public void SetTurnLeft(double degrees)
    {
      __botInternals.SetTurnLeft(degrees);
    }

    /// <inheritdoc/>
    public void TurnLeft(double degrees)
    {
      __botInternals.TurnLeft(degrees);
    }

    /// <inheritdoc/>
    public void SetTurnRight(double degrees)
    {
      SetTurnLeft(-degrees);
    }

    /// <inheritdoc/>
    public void TurnRight(double degrees)
    {
      TurnLeft(-degrees);
    }

    /// <inheritdoc/>
    public double TurnRemaining => __botInternals.TurnRemaining;

    /// <inheritdoc/>
    public void SetTurnGunLeft(double degrees)
    {
      __botInternals.SetTurnGunLeft(degrees);
    }

    /// <inheritdoc/>
    public void TurnGunLeft(double degrees)
    {
      __botInternals.TurnGunLeft(degrees);
    }

    /// <inheritdoc/>
    public void SetTurnGunRight(double degrees)
    {
      SetTurnGunLeft(-degrees);
    }

    /// <inheritdoc/>
    public void TurnGunRight(double degrees)
    {
      TurnGunLeft(-degrees);
    }

    /// <inheritdoc/>
    public double GunTurnRemaining => __botInternals.GunTurnRemaining;

    /// <inheritdoc/>
    public void SetTurnRadarLeft(double degrees)
    {
      __botInternals.SetTurnRadarLeft(degrees);
    }

    /// <inheritdoc/>
    public void TurnRadarLeft(double degrees)
    {
      __botInternals.TurnRadarLeft(degrees);
    }

    /// <inheritdoc/>
    public void SetTurnRadarRight(double degrees)
    {
      SetTurnRadarLeft(-degrees);
    }

    /// <inheritdoc/>
    public void TurnRadarRight(double degrees)
    {
      __botInternals.TurnRadarLeft(-degrees);
    }

    /// <inheritdoc/>
    public double RadarTurnRemaining => __botInternals.RadarTurnRemaining;

    /// <inheritdoc/>
    public void Fire(double firepower)
    {
      __botInternals.Fire(firepower);
    }

    /// <inheritdoc/>
    public void Stop()
    {
      __botInternals.Stop();
    }

    /// <inheritdoc/>
    public void Resume()
    {
      __botInternals.Resume();
    }

    /// <inheritdoc/>
    public void Scan()
    {
      __botInternals.Scan();
    }

    /// <inheritdoc/>
    public void WaitFor(Condition condition)
    {
      __botInternals.WaitFor(condition);
    }

    /// <inheritdoc/>
    public void SetTargetSpeed(double targetSpeed)
    {
      __botInternals.SetTargetSpeed(targetSpeed);
    }
  }
}

File: bot-api/typescript/src/Bot.ts
import { IBot } from "./IBot.js";
import { BaseBot } from "./BaseBot.js";
import { BotInfo } from "./BotInfo.js";
import { Condition } from "./events/Condition.js";
import { BotInternals } from "./internal/BotInternals.js";

/**
 * Abstract bot class that provides convenient methods for movement, turning, and firing.
 * Most bots should extend this class rather than BaseBot.
 */
export abstract class Bot extends BaseBot implements IBot {
  private readonly _botInternals: BotInternals;

  constructor();
  constructor(botInfo: BotInfo);
  constructor(botInfo: BotInfo, serverUrl: string);
  constructor(botInfo: BotInfo, serverUrl: string, serverSecret: string);
  constructor(botInfo?: BotInfo, serverUrl?: string, serverSecret?: string) {
    if (botInfo === undefined) {
      super();
    } else if (serverUrl === undefined) {
      super(botInfo);
    } else if (serverSecret === undefined) {
      super(botInfo, serverUrl);
    } else {
      super(botInfo, serverUrl, serverSecret);
    }
    this._botInternals = new BotInternals(this, this._internals);
  }

  // Override rate/speed setters to go through BotInternals (tracks remaining values)
  override setTurnRate(turnRate: number): void { this._botInternals.setTurnRate(turnRate); }
  override setGunTurnRate(gunTurnRate: number): void { this._botInternals.setGunTurnRate(gunTurnRate); }
  override setRadarTurnRate(radarTurnRate: number): void { this._botInternals.setRadarTurnRate(radarTurnRate); }
  override setTargetSpeed(targetSpeed: number): void { this._botInternals.setTargetSpeed(targetSpeed); }

  isRunning(): boolean { return this._internals.isRunning(); }

  run(): void { /* override in subclass */ }

  // Setter movement (non-blocking)
  setForward(distance: number): void { this._botInternals.setForward(distance); }
  setBack(distance: number): void { this._botInternals.setForward(-distance); }
  setTurnLeft(degrees: number): void { this._botInternals.setTurnLeft(degrees); }
  setTurnRight(degrees: number): void { this._botInternals.setTurnLeft(-degrees); }
  setTurnGunLeft(degrees: number): void { this._botInternals.setTurnGunLeft(degrees); }
  setTurnGunRight(degrees: number): void { this._botInternals.setTurnGunLeft(-degrees); }
  setTurnRadarLeft(degrees: number): void { this._botInternals.setTurnRadarLeft(degrees); }
  setTurnRadarRight(degrees: number): void { this._botInternals.setTurnRadarLeft(-degrees); }

  // Blocking movement
  forward(distance: number): void { this._botInternals.forward(distance); }
  back(distance: number): void { this._botInternals.forward(-distance); }
  turnLeft(degrees: number): void { this._botInternals.turnLeft(degrees); }
  turnRight(degrees: number): void { this._botInternals.turnLeft(-degrees); }
  turnGunLeft(degrees: number): void { this._botInternals.turnGunLeft(degrees); }
  turnGunRight(degrees: number): void { this._botInternals.turnGunLeft(-degrees); }
  turnRadarLeft(degrees: number): void { this._botInternals.turnRadarLeft(degrees); }
  turnRadarRight(degrees: number): void { this._botInternals.turnRadarLeft(-degrees); }

  // Remaining getters
  getDistanceRemaining(): number { return this._botInternals.distanceRemaining; }
  getTurnRemaining(): number { return this._botInternals.turnRemaining; }
  getGunTurnRemaining(): number { return this._botInternals.gunTurnRemaining; }
  getRadarTurnRemaining(): number { return this._botInternals.radarTurnRemaining; }

  // Blocking actions
  fire(firepower: number): void { this._botInternals.fire(firepower); }

  override setStop(): void;
  override setStop(overwrite: boolean): void;
  override setStop(overwrite?: boolean): void { this._internals.setStop(overwrite); }

  stop(): void;
  stop(overwrite: boolean): void;
  stop(overwrite?: boolean): void { this._botInternals.stop(overwrite); }

  resume(): void { this._botInternals.resume(); }
  rescan(): void { this._botInternals.rescan(); }
  waitFor(condition: Condition): void { this._botInternals.waitFor(() => condition.test()); }
}
